common/cuda_hip/solver/batch_tridiagonal_solver_kernels.hpp.inc

```diff
@@ -0,0 +1,878 @@
+        // coalesced accesses while writing data
+        c_batch_entry[my_row_idx] = my_c;
+        d_batch_entry[my_row_idx] = my_d;
+    }
+}
+
+
+template <typename ValueType, typename Group>
+__device__ __forceinline__ void backward_substitution(
+    Group& subwarp_grp, const int nrows, const int my_row_idx,
+    const int tile_size, const int final_group_size,
+    ValueType& x_first_higher_group,
+    const ValueType* const __restrict__ c_batch_entry,
+    const ValueType* const __restrict__ d_batch_entry,
+    ValueType* const __restrict__ x_batch_entry)
+{
+    const int lane = subwarp_grp.thread_rank();
+    const int grp_idx = lane / final_group_size;
+    const int num_grps = tile_size / final_group_size;
+
+    ValueType my_c, my_d, my_x;
+    if (my_row_idx < nrows) {
+        my_c = c_batch_entry[my_row_idx];
+        my_d = d_batch_entry[my_row_idx];
+        // coalseced accesses while reading
+    }
+
+    for (int i = num_grps - 1; i >= 0; i--) {
+        if (grp_idx == i) {
+            my_x = my_d - x_first_higher_group * my_c;
+        }
+
+        subwarp_grp.sync();
+        const int target_lane = i * final_group_size;
+        x_first_higher_group = subwarp_grp.shfl(my_x, target_lane);
+    }
+
+    if (my_row_idx < nrows) {
+        x_batch_entry[my_row_idx] = my_x;
+        // coalseced accessed while writing
+    }
+}
+
+}  // namespace WM_pGE_approach1
+
+template <int subwarp_size, typename ValueType>
+__global__ __launch_bounds__(default_block_size) void WM_pGE_kernel_approach_1(
+    const int num_WM_steps, const size_type nbatch, const int nrows,
+    const ValueType* const __restrict__ a,
+    const ValueType* const __restrict__ b, ValueType* const __restrict__ c,
+    ValueType* const __restrict__ d, ValueType* const __restrict__ x)
+{
+    auto subwarpgrp =
+        group::tiled_partition<subwarp_size>(group::this_thread_block());
+    const int subgrpwarp_id_in_grid =
+        thread::get_subwarp_id_flat<subwarp_size, int>();
+    const int total_num_subwarp_grps_in_grid =
+        thread::get_subwarp_num_flat<subwarp_size, int>();
+    const int id_within_warp = subwarpgrp.thread_rank();
+
+    // a subwarp per matrix in the batch
+    for (size_type batch_idx = subgrpwarp_id_in_grid; batch_idx < nbatch;
+         batch_idx += total_num_subwarp_grps_in_grid) {
+        // Approach: a thread in the subwarp handles one row of the matrix or to
+        // be precise, a row in the matrix tile
+        const auto tile_size = subwarp_size;
+        const auto num_tiles = ceildiv(nrows, tile_size);
+        const bool is_last_tile_similar = ((nrows % tile_size) == 0);
+        const int final_group_size = pow(2, num_WM_steps);
+
+        if (final_group_size > tile_size) {
+            printf(
+                "\n Number of WM steps ans subwarp size should satisfy the "
+                "inequality: pow(2, WM_steps) <=  tile_size, where tile_size = "
+                "subwarp_size");
```

Comment:
I think condition requirement should be mentioned and maybe even verified when launching the kernel. So, we shouldn't be checking this condition here.